isbnbot/normalize_isbns.py

```diff
@@ -128,66 +76,66 @@ def run(self, dump_filepath: str) -> None:
                         self.save(lambda: edition.save(comment=comment))
 
-    def save(self, save_fn):
-        """Modify default save behavior based on 'limit' and 'dry_run' parameters"""
-        if not self.dry_run:
-            save_fn()
-        else:
-            self.logger.info("Modification not made because dry_run is True")
-        self.changed += 1
-        if self.limit and self.changed >= self.limit:
-            self.logger.info("Modification limit reached. Exiting script.")
-            sys.exit()
-
-
-def str2bool(value):
-    if isinstance(value, bool):
-        return value
-    if value.lower() in ("yes", "true", "t", "y", "1"):
-        return True
-    elif value.lower() in ("no", "false", "f", "n", "0"):
-        return False
-    else:
-        raise argparse.ArgumentTypeError("Boolean value expected.")
-
 
-def dedupe(input: list) -> list:
+def dedupe(input_list: list) -> list:
     """Remove duplicate elements in a list and return the new list"""
     output = list()
-    for i in input:
+    for i in input_list:
         if i not in output:
             output.append(i)
     return output
 
 
+def parse_isbns(string: str) -> list:
+    """Find isbns in a string on the assumption that if you strip all non-isbn
+    characters and all that is left is valid isbns then it's unlikely to be random chance
+    if that doesn't work it will use a reasonably strict regex to look for an isbn13
+    formatted string within the full text, which should help when other numbers are present"""
+
+    # pass one: strip all non isbn characters, and see if what remains looks like an ISBN
+    isbnchars = [c for c in string if c in "0123456789Xx"]
+    if len(isbnchars) < 10:
+        return []
+
+    isbnchars = "".join(isbnchars).upper()
+    # X is tricky, but can only appear at the end of an isbn10 so remove if not where expected
+    x_idx = isbnchars.find("X")
+    if (x_idx + 1) % 10 != 0 and x_idx + 1 != 23:
+        isbnchars = isbnchars.replace("X", "")
+
+    if len(isbnchars) % 10 == 0:
+        if all(isbnlib.is_isbn10(isbn) for isbn in split(isbnchars, 10)):
+            return split(isbnchars, 10)
+    elif len(isbnchars) % 13 == 0:
+        if all(isbnlib.is_isbn13(isbn) for isbn in split(isbnchars, 13)):
+            return split(isbnchars, 13)
+    elif (
+        len(isbnchars) == 23
+        and isbnlib.is_isbn13(isbnchars[:13])
+        and isbnlib.is_isbn10(isbnchars[13:])
+    ):
+        return [isbnchars[:13], isbnchars[13:]]
+    elif (
+        len(isbnchars) == 23
+        and isbnlib.is_isbn10(isbnchars[:10])
+        and isbnlib.is_isbn13(isbnchars[10:])
+    ):
+        return [isbnchars[:10], isbnchars[10:]]
+
+    # if we get this far then  we have 10+ string, but it's not a ISBN 10, 13 or a basic combination of the two. This is
+    # beyond the scope of this bot job
+    return []
+
+
+def split(string, length):
+    return list(string[i : i + length] for i in range(0, len(string), length))
```

Comment:
```suggestion
[`str.split()`](https://docs.python.org/3.3/library/stdtypes.html#str.split) already exists so chose a slightly different function name to avoid confusing the reader, and perhaps move this function up above where it is first used.  Add type hints, doctests, and return tuples because they take less memory than lists and signal to the reader that they should not be modified at runtime...
```suggestion
def split_at(string: str, length: int) -> tuple[str]:
    """
    >>> for i in range(1, 9):
    ...     split("example", i)
    ('e', 'x', 'a', 'm', 'p', 'l', 'e')
    ('ex', 'am', 'pl', 'e')
    ('exa', 'mpl', 'e')
    ('exam', 'ple')
    ('examp', 'le')
    ('exampl', 'e')
    ('example')
    ('example')
    """
    return tuple(string[i : i + length] for i in range(0, len(string), length))
```

Reply:
Actually, it would be a fair amount of work to enable pytest to run doctests on this repo (see  #252) so these should be converted to normal pytests.